File: aws/openscap_securityhub/lambda/lambda.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/cloudwatch"
	"github.com/aws/aws-sdk-go/service/dynamodb"
	"github.com/aws/aws-sdk-go/service/s3"
	"github.com/aws/aws-sdk-go/service/securityhub"
	"github.com/aws/aws-sdk-go/service/ssm"
	"github.com/aws/aws-sdk-go/service/ssm/ssmiface"
)

var (
	s3Client         *s3.S3
	cloudWatchClient *cloudwatch.CloudWatch
	dynamoDBClient   *dynamodb.DynamoDB
	securityHubClient *securityhub.SecurityHub
	ssmClient        ssmiface.SSMAPI
)

func init() {
	sess := session.Must(session.NewSessionWithOptions(session.Options{
		SharedConfigState: session.SharedConfigEnable,
	}))
	s3Client = s3.New(sess)
	cloudWatchClient = cloudwatch.New(sess)
	dynamoDBClient = dynamodb.New(sess)
	securityHubClient = securityhub.New(sess)
	ssmClient = ssm.New(sess)
}

func lambdaHandler(ctx context.Context, event events.S3Event) error {
	bucketName := event.Records[0].S3.Bucket.Name
	fileKey := event.Records[0].S3.Object.Key
	awsAccountID := strings.Split(ctx.InvokedFunctionArn, ":")[4]
	region := strings.Split(ctx.InvokedFunctionArn, ":")[3]

	instanceID := strings.Split(fileKey, "/")[0]

	obj, err := s3Client.GetObject(&s3.GetObjectInput{
		Bucket: aws.String(bucketName),
		Key:    aws.String(fileKey),
	})
	if err != nil {
		return err
	}

	useSecurityHub, err := getParameter("/SCAPTesting/EnableSecurityHub")
	if err != nil {
		return err
	}

	root := ET.fromstring(obj.Body)
	testResult := root.find(".//{http://checklists.nist.gov/xccdf/1.2}TestResult")
	testVersion := testResult.Attr["version"]

	high, medium, low, unknown := 0, 0, 0, 0

	ignoreList, err := getIgnoreList()
	if err != nil {
		return err
	}

	var dynamoDBItems []map[string]*dynamodb.AttributeValue
	var securityHubFindings []*securityhub.AWSSecurityFinding

	for _, item := range testResult {
		testID := item.Attr["idref"]
		if strings.Contains(testID, ".") {
			testID = testID[strings.LastIndex(testID, ".")+1:]
		}

		if !contains(ignoreList, testID) {
			if item.Find(".//{http://checklists.nist.gov/xccdf/1.2}result").Text() == "fail" {
				saveToDynamoDB(&dynamoDBItems, instanceID, item, bucketName, fileKey)
				if useSecurityHub == "yes" && contains([]string{"high", "medium", "low"}, item.Attr["severity"]) {
					err := pushToSecurityHub(&securityHubFindings, root, instanceID, item, region, awsAccountID, testVersion, bucketName, fileKey)
					if err != nil {
						useSecurityHub = "no"
						fmt.Printf("SecurityHub is not enabled b: %v\n", err)
					}
				}
				switch item.Attr["severity"] {
				case "high":
					high++
				case "medium":
					medium++
				case "low":
					low++
				case "unknown":
					unknown++
				}
			}
		}
	}

	sendMetric(high, "SCAP High Finding", instanceID)
	sendMetric(medium, "SCAP Medium Finding", instanceID)
	sendMetric(low, "SCAP Low Finding", instanceID)

	batchWriteToDynamoDB(&dynamoDBItems)

	if useSecurityHub == "yes" {
		err := batchSubmitToSecurityHub(securityHubFindings)
		if err != nil {
			fmt.Printf("SecurityHub is not enabled a: %v\n", err)
		}
	}

	return nil
}

func saveToDynamoDB(dynamoDBItems *[]map[string]*dynamodb.AttributeValue, instanceID, item, bucketName, fileKey string) {
	dynamoDBItems = append(dynamoDBItems, map[string]*dynamodb.AttributeValue{
		"InstanceId":      {S: aws.String(instanceID)},
		"SCAP_Rule_Name":  {S: aws.String(item.Attr["idref"])},
		"time":            {S: aws.String(item.Attr["time"])},
		"severity":        {S: aws.String(item.Attr["severity"])},
		"result":          {S: aws.String(item.Find(".//{http://checklists.nist.gov/xccdf/1.2}result").Text())},
		"report_url":      {S: aws.String(fmt.Sprintf("s3://%s/%s", bucketName, strings.ReplaceAll(fileKey, ".xml", ".html")))},
	})
}

func sendMetric(value int, title, instanceID string) {
	_, err := cloudWatchClient.PutMetricData(&cloudwatch.PutMetricDataInput{
		Namespace: aws.String("Compliance"),
		MetricData: []*cloudwatch.MetricDatum{
			{
				MetricName: aws.String(title),
				Dimensions: []*cloudwatch.Dimension{
					{
						Name:  aws.String("InstanceId"),
						Value: aws.String(instanceID),
					},
				},
				Value: aws.Float64(float64(value)),
			},
		},
	})
	if err != nil {
		fmt.Printf("Error sending metric: %v\n", err)
	}
}

func getIgnoreList() ([]string, error) {
	input := &dynamodb.ScanInput{
		TableName: aws.String("SCAP_Scan_Ignore_List"),
	}
	result, err := dynamoDBClient.Scan(input)
	if err != nil {
		return nil, err
	}
	var ignoreList []string
	for _, item := range result.Items {
		ignoreList = append(ignoreList, aws.StringValue(item["SCAP_Rule_Name"].S))
	}
	return ignoreList, nil
}

func pushToSecurityHub(securityHubFindings *[]*securityhub.AWSSecurityFinding, root *ET.Element, instanceID, item, region, awsAccountID, testVersion, bucketName, fileKey string) error {
	rule := root.Find(fmt.Sprintf(".//{http://checklists.nist.gov/xccdf/1.2}Rule[@id='%s']", item.Attr["idref"]))
	profile := root.Find(".//{http://checklists.nist.gov/xccdf/1.2}Profile[@id='xccdf_org.ssgproject.content_profile_stig']")

	timeStr := item.Attr["time"]
	if strings.Index(timeStr, "+") != -1 {
		timeStr = timeStr[:strings.LastIndex(timeStr, "+")]
	}
	timeStr += ".000Z"

	*securityHubFindings = append(*securityHubFindings, &securityhub.AWSSecurityFinding{
		SchemaVersion: aws.String("2018-10-08"),
		Id:            aws.String(item.Attr["idref"] + "_" + fileKey),
		ProductArn:    aws.String(fmt.Sprintf("arn:aws:securityhub:%s:%s:product/%s/default", region, awsAccountID, awsAccountID)),
		GeneratorId:   aws.String(fmt.Sprintf("OpenSCAP %s", item.Attr["idref"])),
		AwsAccountId:  aws.String(awsAccountID),
		Types: []*string{
			aws.String("Software and Configuration Checks"),
		},
		FirstObservedAt: aws.String(timeStr),
		LastObservedAt:  aws.String(timeStr),
		CreatedAt:       aws.String(timeStr),
		UpdatedAt:       aws.String(timeStr),
		Severity: &securityhub.Severity{
			Label: aws.String(strings.ToUpper(item.Attr["severity"])),
		},
		Title: aws.String(rule.Find(".//{http://checklists.nist.gov/xccdf/1.2}title").Text()),
		Description: aws.String(fmt.Sprintf("%s ", rule.Find(".//{http://checklists.nist.gov/xccdf/1.2}description").Text())),
		Remediation: &securityhub.Remediation{
			Recommendation: &securityhub.Recommendation{
				Text: aws.String(fmt.Sprintf("For remediation please see: s3://%s/%s", bucketName, strings.ReplaceAll(fileKey, ".xml", ".html"))),
			},
		},
		ProductFields: map[string]*string{
			"ProviderName":    aws.String(profile.Find(".//{http://checklists.nist.gov/xccdf/1.2}title").Text()),
			"ProviderVersion": aws.String(testVersion),
		},
		Resources: []*securityhub.Resource{
			{
				Type:   aws.String("AwsEc2Instance"),
				Id:     aws.String(instanceID),
				Region: aws.String(region),
			},
		},
		Compliance: &securityhub.Compliance{
			Status: aws.String("FAILED"),
		},
		WorkflowState: aws.String("NEW"),
		Workflow: &securityhub.Workflow{
			Status: aws.String("NEW"),
		},
	})
}
}

func getParameter(paramName string) (string, error) {
input := &ssm.GetParameterInput{
	Name:           aws.String(paramName),
	WithDecryption: aws.Bool(true),
}
result, err := ssmClient.GetParameter(input)
if err != nil {
	return "", err
}
return aws.StringValue(result.Parameter.Value), nil
}

func contains(s []string, e string) bool {
for _, a := range s {
	if a == e {
		return true
	}
}
return false
}

func batchWriteToDynamoDB(items *[]map[string]*dynamodb.AttributeValue) {
tableName := "SCAP_Scan_Results"
writeRequests := make([]*dynamodb.WriteRequest, len(*items))
for i, item := range *items {
	writeRequests[i] = &dynamodb.WriteRequest{
		PutRequest: &dynamodb.PutRequest{
			Item: item,
		},
	}
}

batchRequest := &dynamodb.BatchWriteItemInput{
	RequestItems: map[string][]*dynamodb.WriteRequest{
		tableName: writeRequests,
	},
}

dynamoDBClient.BatchWriteItem(batchRequest)
}

func batchSubmitToSecurityHub(findings []*securityhub.AWSSecurityFinding) error {
batchSize := 100
for i := 0; i < len(findings); i += batchSize {
	end := i + batchSize
	if end > len(findings) {
		end = len(findings)
	}
	chunk := findings[i:end]

	batchRequest := &securityhub.BatchImportFindingsInput{
		Findings: chunk,
	}

	_, err := securityHubClient.BatchImportFindings(batchRequest)
	if err != nil {
		return err
	}
}
return nil
}

func main() {
lambda.Start(lambdaHandler)
}

File: aws/openscap_securityhub/test-lambda/lambda.py
import json
import boto3
import datetime
from datetime import date
import xml.etree.ElementTree as ET

s3 = boto3.client('s3')
cloudWatch = boto3.client('cloudwatch')
dynamodb = boto3.resource('dynamodb')
securityHub = boto3.client('securityhub')
ssmClient = boto3.client('ssm')


def main():
    
    # get the bucket name so that we can get the file from s3
    bucket_name = "testingcftmarch2024"
    file_key = "scap-results.xml"
    aws_account_id = "344594102751"
    region = "us-east-1"
    
    #get the instance id from the s3 path
    instanceId = file_key.split('/')[0]
    
    # get the object
    obj = s3.get_object(Bucket=bucket_name, Key=file_key)
    
    # r = requests.get('https://dl.dod.cyber.mil/wp-content/uploads/stigs/zip/U_RHEL_9_V1R2_STIG.zip')

    # Get parameter for using Security Hub
    useSecurityHub = ssmClient.get_parameter(Name='/SCAPTesting/EnableSecurityHub')['Parameter']['Value']
    
    # parse the XML from s3
    root = ET.fromstring(obj['Body'].read())
    
    # get the resutls node from the xml
    testResult = root.find(".//{http://checklists.nist.gov/xccdf/1.2}TestResult")
    testVersion = testResult.attrib.get("version") 
    
    # setup counts for cloudwatch metrics
    high=0
    medium=0
    low=0
    unknown=0
    
    
    # load the ignore list from DynamoDB
    ignoreList = getIgnoreList()
    
    # setup arrays to hold the findings so we can do batch inserts
    dynamoDbItems = []
    securityHubFindings = []
    
    # iterate through each result item
    for item in testResult: 
        testId = str(item.attrib.get("idref"))

        # We need to normalize the rule name here to check agains the
        # ignore list
        if '.' in testId:
            testId = testId[testId.rindex('.')+1:len(testId)]

        # if we are not ignoring the result, them count it and store in DynamoDB
        if testId not in ignoreList:
            if(item.findtext('{http://checklists.nist.gov/xccdf/1.2}result') == "fail"):
                saveToDynamoDB(dynamoDbItems, instanceId, item, bucket_name, file_key)
                pushToSecurityHub(securityHubFindings,root, instanceId, item, region, aws_account_id, testVersion, bucket_name, file_key)
                # if useSecurityHub == "true" and item.attrib.get("severity") in ["high","medium","low"]:
                #     try:
                #         pushToSecurityHub(securityHubFindings,root, instanceId, item, region, aws_account_id, testVersion, bucket_name, file_key)
                #     except Exception as e:
                #         useSecurityHub = "false"
                #         print("SecurityHub is not enabled b: " + str(e))
                if(item.attrib.get("severity") == "high"):
                    high+=1
                elif(item.attrib.get("severity") == "medium"):
                    medium+=1
                elif(item.attrib.get("severity") == "low"):
                    low+=1
                elif(item.attrib.get("severity") == "unknown"):
                    unknown+=1
            
    # Send metrics to cloudwatch for alerting        
    sendMetic(high, 'SCAP High Finding', instanceId)
    sendMetic(medium, 'SCAP Medium Finding', instanceId)
    sendMetic(low, 'SCAP Low Finding', instanceId)
    # Batch write all findings to DynamoDB
    table = dynamodb.Table('SCAP_Scan_Results')
    with table.batch_writer() as batch:
        for item in dynamoDbItems:
            batch.put_item(
                Item = item
            )
    
    # if Security Hub is enabled, send the results in batches of 100
    if useSecurityHub == "true":
        myfindings = securityHubFindings
        try:
            findingsLeft = True
            startIndex = 0
            stopIndex = len(myfindings)

            # Loop through the findings sending 100 at a time to Security Hub
            while findingsLeft:
                stopIndex = startIndex + 100
                if stopIndex > len(securityHubFindings):
                    stopIndex = len(securityHubFindings)
                    findingsLeft = False
                else:
                    stopIndex = 100
                myfindings = securityHubFindings[startIndex:stopIndex]
                # submit the finding to Security Hub
                result = securityHub.batch_import_findings(Findings = myfindings)
                startIndex = startIndex + 100

                # print results to CloudWatch
                print(result)
        except Exception as e:
            print("SecurityHub is not enabled a: " + str(e))
            

# Saves the results to DynamoDB   
def saveToDynamoDB(dynamoDbItems, instanceId, item, bucket_name, file_key):
    #table = dynamodb.Table('SCAP_Scan_Results')
    #table.put_item(
    dynamoDbItems.append({
            'InstanceId': instanceId,
            'SCAP_Rule_Name': item.attrib.get("idref"),
            'time': item.attrib.get("time"), 
            'severity':  item.attrib.get("severity"),
            'result': item.findtext('{http://checklists.nist.gov/xccdf/1.2}result'),
            'report_url': 's3://'+ bucket_name + "/" + file_key.replace('.xml', '.html')
            }
    )


# method for creating the metrics
def sendMetic(value, title, instanceId):
    cloudWatch.put_metric_data(
        Namespace='Compliance',
        MetricData=[
            {
                'MetricName': title,
                'Dimensions': [
                    {
                        'Name': 'InstanceId',
                        'Value': instanceId
                    },
                ],
                'Value': value
            }
        ]
    )

# fetches the ignore list from DynamoDB    
def getIgnoreList():
    table = dynamodb.Table('SCAP_Scan_Ignore_List')
    #if you list is really long this could fail as it will pagonate
    response = table.scan()
    list = response['Items']
    returnList = []
    for item in list:
        returnList.append(item['SCAP_Rule_Name'])
    return returnList
    
def pushToSecurityHub(securityHubFindings, root, instanceId, item, region, aws_account_id, testVersion, bucket_name, file_key):
    rule = root.find(".//{http://checklists.nist.gov/xccdf/1.2}Rule[@id='" + item.attrib.get("idref") + "']")
    profile = root.find('.//{http://checklists.nist.gov/xccdf/1.2}Profile[@id="xccdf_org.ssgproject.content_profile_stig"]')

    # fix the time format from OpenSCAP to Security Hub
    time = item.attrib.get("time")
    if time.find('+') != -1:
        time = time[:time.rindex('+')]
    time =  time + ".000Z"
    
    securityHubFindings.append(
            {
                'SchemaVersion': '2018-10-08',
                'Id': item.attrib.get("idref") + "_" + file_key,
                'ProductArn': 'arn:aws:securityhub:' + region + ':'+ aws_account_id +':product/' + aws_account_id + '/default',
                'GeneratorId': 'OpenSCAP ' + item.attrib.get("idref"),
                'AwsAccountId': aws_account_id,
                'Types': [
                    'Software and Configuration Checks',
                ],
                'FirstObservedAt': time,
                'LastObservedAt': time,
                'CreatedAt': time,
                'UpdatedAt': time,
                'Severity': {
                    'Label': item.attrib.get("severity").upper()
                },
                'Title': rule.findtext('{http://checklists.nist.gov/xccdf/1.2}title'),
                'Description': str(rule.findtext('{http://checklists.nist.gov/xccdf/1.2}description')) + " ",
                'Remediation': {
                    'Recommendation': {
                        'Text': 'For remediation please see: s3://'+ bucket_name + '/' + file_key.replace('.xml', '.html')
                    }
                },
                'ProductFields': {
                    "ProviderName": str(rule.findtext('{http://checklists.nist.gov/xccdf/1.2}title')) + " ",
                    "ProviderVersion": testVersion
                },
                'Resources': [
                    {
                        'Type': 'AwsEc2Instance',
                        'Id': instanceId,
                        'Region': region
                    },
                ],
                'Compliance': {
                    'Status': 'FAILED'
                },
                'WorkflowState': 'NEW',
                'Workflow': {
                    'Status': 'NEW'
                }
            })
    
if __name__ == "__main__":
    main()